File: KMACTF 2024/ZKP for QNR/python/main.py
import os
import random
from typing import List, Tuple
from dataclasses import dataclass
from pathlib import Path

# Utility functions (equivalent to the Rust module `utils`)
def bytes_to_bigint(bytes_data: bytes) -> int:
    """Convert a byte array to an integer."""
    return int.from_bytes(bytes_data, byteorder='big')

def rand_below(n: int) -> int:
    """Generate a random integer less than n."""
    bit_length = n.bit_length()
    while True:
        y = random.getrandbits(bit_length)
        if y < n:
            return y

def convert_rp_to_string(v: List['Rp']) -> List[str]:
    """Convert a list of Rp objects to a list of strings."""
    normal_numbers = []
    for item in v:
        if isinstance(item, RpTuple):
            normal_numbers.append(f"({item.a}, {item.b})")
        elif isinstance(item, RpNumber):
            normal_numbers.append(f"{item.n}")
    return normal_numbers

# Classes equivalent to the Rust modules

@dataclass
class Prover:
    x: int
    y: int
    num_rounds: int

    def gen_list_i(self) -> List[int]:
        """Generate a list of random bits (0 or 1) of length num_rounds."""
        return [random.randint(0, 1) for _ in range(self.num_rounds)]

@dataclass
class Verifier:
    x: int
    y: int
    rs: List[List[Tuple[int, int]]] = field(default_factory=list)
    pairs: List[List[Tuple[int, int]]] = field(default_factory=list)
    num_rounds: int = 0

    def gen_w(self, bit: int, r: int) -> int:
        """Generate w based on the bit value and r."""
        if bit == 0:
            return pow(r, 2, self.x)
        else:
            return (pow(r, 2, self.x) * self.y) % self.x

    def gen_pairs(self) -> List[Tuple[int, int]]:
        """Generate pairs of values."""
        rr = []
        pair = []
        for _ in range(self.num_rounds):
            r1 = rand_below(self.x)
            r2 = rand_below(self.x)
            a = pow(r1, 2, self.x)
            b = (pow(r2, 2, self.x) * self.y) % self.x

            pair.append((a, b))
            rr.append((r1, r2))
        self.rs.append(rr)
        self.pairs.append(pair)
        return pair

    def respond(self, bit: int, list_i: List[int], r: int) -> List['Rp']:
        """Generate a list of responses based on the bit, list_i, and r."""
        v = []
        for i, a in enumerate(list_i):
            if a == 0:
                vj = self.rs[-1][i]
                v.append(RpTuple(vj[0], vj[1]))
            elif a == 1:
                if bit == 0:
                    vj = (r * self.rs[-1][i][0]) % self.x
                elif bit == 1:
                    vj = (r * self.rs[-1][i][1] * self.y) % self.x
                v.append(RpNumber(vj))
            else:
                raise ValueError("Invalid input")
        return v

    def get_r(self) -> int:
        """Get a random value below x."""
        return rand_below(self.x)

@dataclass
class RpTuple:
    a: int
    b: int

@dataclass
class RpNumber:
    n: int

def main():
    flag = b"KMACTF{*******************************************}"
    num_flag = bytes_to_bigint(flag)

    num_rounds = num_flag.bit_length()

    folder_name = "Log"

    # Create the folder if it doesn't exist
    Path(folder_name).mkdir(parents=True, exist_ok=True)

    print(f"Number of rounds: {num_rounds}")

    x = int("106276637345585586395178695555113419125706596151484787339368729136766801222943")
    y = int("67502870359608496464376733754660348616157530226832182619029429292243849029379")

    prover = Prover(x, y, num_rounds)
    verifier = Verifier(x, y, num_rounds)

    for i in range(num_rounds):
        file_name = f"{folder_name}/output_{i}.txt"
        with open(file_name, "w") as file:
            file.write(f"Round {i}\n")

            bit = (num_flag >> i) & 1
            r = verifier.get_r()

            # give the prover the pair and r
            w = verifier.gen_w(bit, r)
            pair = verifier.gen_pairs()

            file.write(f"w = {w}\n")

            file.write("Pair = \n")
            for p in pair:
                file.write(f"({p[0]}, {p[1]})\n")

            list_i = prover.gen_list_i()
            file.write(f"list_i = {list_i}\n")

            v = verifier.respond(bit, list_i, r)

            v_str = convert_rp_to_string(v)

            file.write("list_of_respond = \n")
            for line in v_str:
                file.write(f"{line}\n")

if __name__ == "__main__":
    main()

File: KMACTF 2024/ZKP for QNR/src/main.rs
use num_bigint::BigUint;
use std::str::FromStr;
pub mod prover;
pub mod utils;
pub mod verifier;
use num_traits::cast::ToPrimitive;
use prover::*;
use std::fs::{self, File};
use std::io::{BufWriter, Write};
use utils::*;
use verifier::Verifier;

fn main() -> std::io::Result<()> {
    let flag = b"KMACTF{*******************************************}";
    let num_flag = bytes_to_biguint(flag);

    let num_rounds = num_flag.bits() as usize;

    // Define the folder name
    let folder_name = "Log";

    // Create the folder if it doesn't exist
    if fs::metadata(folder_name).is_err() {
        fs::create_dir(folder_name)?;
    }

    println!("Number of rounds: {}", num_rounds);

    let x = BigUint::from_str(
        "106276637345585586395178695555113419125706596151484787339368729136766801222943",
    )
    .unwrap();
    let y = BigUint::from_str(
        "67502870359608496464376733754660348616157530226832182619029429292243849029379",
    )
    .unwrap();

    let prover = Prover::new(x.clone(), y.clone(), num_rounds);
    let mut verifier = Verifier::new(x.clone(), y.clone(), num_rounds);

    for i in 0..num_rounds {
        let file_name = format!("{}/output_{}.txt", folder_name, i);
        let file = File::create(&file_name)?;

        let mut writer = BufWriter::new(file);

        writeln!(writer, "Round {}", i)?;

        let bit = ((&num_flag >> i) & BigUint::from(1_u32))
            .to_u32()
            .expect("Can not do this");

        let r = verifier.get_r();

        // give the prover the pair and r
        let w = verifier.gen_w(bit as i32, &r);
        let pair = verifier.gen_pairs();

        writeln!(writer, "w = {}", w)?;

        writeln!(writer, "Pair = ")?;
        for p in pair {
            writeln!(writer, "({}, {})", p.0, p.1)?;
        }

        let list_i = prover.gen_list_i();
        writeln!(writer, "list_i = {:?}", list_i)?;

        let v = verifier.respond(&bit, list_i, &r);

        let v = convert_rp_to_string(&v);

        writeln!(writer, "list_of_respond = ")?;
        for line in v {
            writeln!(writer, "{}", line)?;
        }

        writer.flush()?;
    }

    Ok(())
}
